Documentation/contributors/design-proposals/operator-validation-library.md

```diff
@@ -53,3 +53,191 @@ There is also a Command Line Tool that serves as an example of how a user can in
+Warning: Optional Field Missing (ObjectMeta.Labels)
+Warning: Optional Field Missing (ObjectMeta.OwnerReferences)
+Warning: Optional Field Missing (ObjectMeta.Initializers)
+Warning: Optional Field Missing (ObjectMeta.Finalizers)
+Warning: Optional Field Missing (ObjectMeta.ClusterName)
+Warning: Optional Field Missing (Spec.CustomResourceDefinitions.Required)
+Warning: Optional Struct Missing (Spec.APIServiceDefinitions)
+Warning: Optional Field Missing (Spec.NativeAPIs)
+Warning: Optional Field Missing (Spec.MinKubeVersion)
+Warning: Optional Field Missing (Spec.Provider.URL)
+Warning: Optional Field Missing (Spec.Replaces)
+Warning: Optional Field Missing (Spec.Annotations)
+csv.yaml is verified.
+```
+
+Omitting `Spec.InstallStrategy.StrategyName`, one of the mandatory fields, yields
+
+```
+Error: Mandatory Field Missing (Spec.InstallStrategy.StrategyName)
+Populate all the mandatory fields missing from csv.yaml file.
+```
+
+in addition to the warnings shown above.
+
+To ignore the warnings, we have `-i` flag available.
+
+```
+$ operator-verify -i csv csv.yaml
+```
+
+`crd` and `package` commands works in a similar way as the `csv`. The only difference being `crd` can also accept a directory containing a bunch of CRDs to validate. The `manifest` command returns a log similar to the one shown above for `csv` for each file in the manifest bundle.
+
+# Library
+
+## Getting Started
+
+The Operator Manifest Verfication library provides APIs for validating both individual yaml files and the manifest bundle as a whole. For **individual yaml files**, it checks for: 
+
+* Data type mismatch
+* Missing mandatory and optional fields
+  
+against OLM's type.
+
+For **manifest bundle**, in addition to verifying the individual operator files, this library checks for the CRDs mentioned in the CSV. It accepts both nested and flattened directory structures containing manifest yamls. The  directory structure is expected to adhere to [manifest format](https://github.com/operator-framework/operator-registry#manifest-format). See `usage` for more information. 
+
+### List of Errors
+
+* Unmarshalling errors like
+  * Data type mismatch 
+  * Incorrect indentation
+  * Inconsistent yaml file structure that can't be converted to JSON for unmarshalling
+* Warning for any missing optional field
+* Error for any missing mandatory field
+* Error for missing CRDs which are mentioned in the CSV
+
+Errors and warnings returned by the API are of type `missingTypeError` and can be used to extract more information. `missingTypeError` is struct type and its implementation in the library is as follows,
+
+```go
+type missingTypeError struct {
+	err         string
+	typeName    string
+	path        string
+	isMandatory bool
+}
+```
+
+For each error/warning, we can check if it's a field or a struct (`typeName`), path of that field in the nested yaml file (`path`), and if the field is mandatory or not (`isMandatory`).
+
+### Versioning
```

Comment:
See above there is a section on open questions. Please include this there.

Reply:
Done.